File: app/src/main/java/com/give3/gizrog/give3/SectionSquadActivity.java
package com.give3.gizrog.give3;

import android.annotation.SuppressLint;
import android.content.Context;
import android.content.Intent;
import android.support.annotation.Nullable;
import android.support.v7.app.AppCompatActivity;
import android.os.Bundle;
import android.view.View;
import android.view.ViewGroup;
import android.widget.AdapterView;
import android.widget.BaseAdapter;
import android.widget.Button;
import android.widget.EditText;
import android.widget.ListView;
import android.widget.Toast;

import java.util.ArrayList;
import java.util.List;
import java.util.Objects;

public class SectionSquadActivity extends AppCompatActivity implements AdapterView.OnItemClickListener {

    Bundle controlBundle;

    List<String> studentsNames = new ArrayList<>();

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_section_squad);

        studentsNames.add("add new student");

        Intent intent = getIntent();
        String title = Objects.requireNonNull(intent.getExtras()).getString("Title");
        studentsNames = intent.getExtras().getStringArrayList("Students");

        ListView listView = findViewById(R.id.listview_section_squad);
        CustomViewAdapter customViewAdapter = new CustomViewAdapter();
        listView.setAdapter(customViewAdapter);
    }

    @Nullable
    @Override
    public Intent getParentActivityIntent() {
        return super.getParentActivityIntent();
    }


    public static Intent makeIntent(Context context, Bundle bundle) {
        return new Intent(context, SectionSquadActivity.class);
    }

    @Override
    public void onItemClick(AdapterView<?> adapterView, View view, int i, long l) {
        Toast.makeText(this, "ADDED", Toast.LENGTH_SHORT).show();
        studentsNames.add("new student");
    }

    class CustomViewAdapter extends BaseAdapter {

        @Override
        public int getCount() {
            return studentsNames.size();
        }

        @Override
        public Object getItem(int i) {
            return studentsNames.get(i);
        }

        @Override
        public long getItemId(int i) {
            return 0;
        }

        @SuppressLint("ViewHolder")
        @Override
        public View getView(int i, View view, final ViewGroup viewGroup) {
            view = getLayoutInflater().inflate(R.layout.custom_listview_item, null);
            EditText editText = view.findViewById(R.id.listview_text);
            editText.setText(studentsNames.get(i));

            Button btn = view.findViewById(R.id.add_listitem);
            btn.setOnClickListener(new View.OnClickListener() {
                @Override
                public void onClick(View view) {
                    studentsNames.add("new student...");
                    viewGroup.refreshDrawableState();
                }
            });
            return view;
        }
    }

}

File: app/src/main/java/com/give3/gizrog/give3/SectionSquadActivity.kt
package com.give3.gizrog.give3

import android.app.Dialog
import android.content.Context
import android.content.Intent
import android.support.v7.app.AppCompatActivity
import android.os.Bundle
import android.view.View
import android.view.WindowManager
import android.widget.*
import com.give3.gizrog.give3.adapters.SectionListAdapter
import com.give3.gizrog.give3.models.Section
import kotlinx.android.synthetic.main.activity_section_squad.*

class SectionSquadActivity : AppCompatActivity(), AdapterView.OnItemClickListener {

    private lateinit var section: Section
    private lateinit var listView: ListView
    private lateinit var adapter: SectionListAdapter
    private var requestCode: Int = 0

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        setContentView(R.layout.activity_section_squad)

        section = intent.getParcelableExtra(KEY_SECTION)
        requestCode = intent.getIntExtra(KEY_REQUEST_CODE, 0)

        initialize()
    }

    private fun initialize() {
        initializeListView()
        initializeAddPersonButton()
        initializeDoneButton()
        setTitleText()
    }

    private fun initializeListView() {
        listView = findViewById(R.id.listview_section_squad)
        adapter = SectionListAdapter(this, section.studentsNames)
        listView.onItemClickListener = this
        listView.adapter = adapter
        adapter.notifyDataSetChanged()
    }

    private fun initializeAddPersonButton() {
        findViewById<Button>(R.id.button_section_squad_add).setOnClickListener {
            section.studentsNames.add("Name...")
            adapter.notifyDataSetChanged()
        }
    }

    private fun initializeDoneButton() {
        findViewById<Button>(R.id.button_section_squad_done).setOnClickListener {
            if(section.studentsNames.isEmpty()) {
                Toast.makeText(applicationContext, "Empty section? Come on...", Toast.LENGTH_LONG).show()
            } else {
                val resultIntent = Intent()
                val bundle = Bundle()
                bundle.putParcelable(KEY_SECTION, section)
                resultIntent.putExtras(bundle)
                setResult(requestCode, resultIntent)
                supportFinishAfterTransition()
            }
        }
    }

    private fun setTitleText() {
        text_section_title.text = section.title
    }

    override fun onItemClick(parent: AdapterView<*>?, view: View?, position: Int, id: Long) {
        val dialog = Dialog(this)
        dialog.setTitle("Edit person")
        dialog.setContentView(R.layout.dialog_edit_person)
        val text = dialog.findViewById<EditText>(R.id.text_listitem_edit_person)
        text.setText(section.studentsNames[position])
        dialog.findViewById<Button>(R.id.button_person_edit_done).setOnClickListener {
            section.studentsNames[position] = text.text.toString()
            adapter.notifyDataSetChanged()
            dialog.dismiss()
        }

        dialog.show()
        dialog.window.setLayout(WindowManager.LayoutParams.MATCH_PARENT, WindowManager.LayoutParams.WRAP_CONTENT)
    }

    companion object {

        const val KEY_SECTION = "Section"
        const val KEY_REQUEST_CODE = "RequestCode"
        const val NEW_SECTION = 100
        const val UPDATE_SECTION = 101

        fun makeIntent(context: Context): Intent {
            return Intent(context, SectionSquadActivity::class.java)
        }
    }

}
